Kerberos.NET/Client/KerberosClient.cs

```diff
@@ -597,23 +599,37 @@ private void ConfigurePreAuth(KerberosCredential credential, KerberosProtocolExc
                         rst.Realm = ResolveKdcTarget(tgtEntry);
                         rst.KdcOptions = ReconcileKdcFlags(rst.KdcOptions, tgtEntry.Flags);
 
-                        serviceTicketCacheEntry = await this.RequestTgs(rst, tgtEntry, cancellation).ConfigureAwait(false);
+                        try
+                        {
+                            (serviceTicketCacheEntry, encKdcRepPart) = await this.RequestDecryptedTgs(
+                                rst,
+                                tgtEntry,
+                                cancellation
+                            ).ConfigureAwait(false);
+                        }
+                        catch (KerberosProtocolException pex)
+                            when (pex.Error?.ErrorCode == KerberosErrorCode.KDC_ERR_S_PRINCIPAL_UNKNOWN)
+                        {
+                            // the SPN was authoritatively not found, but that could be because it belongs to another domain
+                            // usually the KDC will automatically refer us to the domain where it thinks it might be instead
+                            // however, sometimes the KDC just doesn't know where we need to go so we check for a hint instead
 
-                        encKdcRepPart = serviceTicketCacheEntry.KdcResponse.EncPart.Decrypt(
-                           serviceTicketCacheEntry.SessionKey.AsKey(),
-                           serviceTicketCacheEntry.SessionKey.Usage,
-                           d => KrbEncTgsRepPart.DecodeApplication(d)
-                       );
+                            logger.LogInformation("Service principal not found. Checking for realm hint. {SPN}", rst.ServicePrincipalName);
 
-                        VerifyNonces(serviceTicketCacheEntry.Nonce, encKdcRepPart.Nonce);
+                            if (this.TryFindRealmHint(rst, out string referral) &&
+                                !string.Equals(tgtCacheName, referral, StringComparison.InvariantCultureIgnoreCase))
+                            {
+                                logger.LogDebug("Realm hint found for {SPN} to {Realm}", rst.ServicePrincipalName, referral);
 
-                        serviceTicketCacheEntry.Flags = encKdcRepPart.Flags;
-                        serviceTicketCacheEntry.AuthTime = encKdcRepPart.AuthTime;
-                        serviceTicketCacheEntry.StartTime = encKdcRepPart.StartTime ?? DateTimeOffset.MinValue;
-                        serviceTicketCacheEntry.EndTime = encKdcRepPart.EndTime;
-                        serviceTicketCacheEntry.RenewTill = encKdcRepPart.RenewTill;
-                        serviceTicketCacheEntry.SessionKey = encKdcRepPart.Key;
-                        serviceTicketCacheEntry.SName = encKdcRepPart.SName;
+                                rst.ServicePrincipalName = referral;
+
+                                continue;
+                            }
+
+                            logger.LogDebug("Realm hint not found");
```

Comment:
should we do another query to verify the mapping?

Reply:
What do you mean?

Reply:
ah. the comment is off. So the logic is:
calls RequestDecryptedTgs(). When it fails, we try to find mapping in configuration to find the correct server/realm. 
Now, I would naively expect that if we do we would talk to that server in order to verify the SPN there....?

Reply:
Ah! It's in a loop, so the next iteration it'll move on to the proposed target and ask again.

1. Dear foo-KDC, I am `xxx` do you have `aaa`?
2. foo-KDC: No, but you can ask `bar-KDC` <returns `xxx-from-foo-KDC`>
4. Dear bar-KDC, I am `xxx-from-foo-KDC` do you have `aaa`?
5. bar-KDC: Yep!

In this case it works nearly the same way but with an extra step added in.

1. Dear foo-KDC, I am `xxx` do you have `aaa`?
2. foo-KDC: No.
3. Dear client-config, do you know where `aaa` is? <---
4. client-config: maybe, try `bar-KDC`
6. Dear foo-KDC, I am `xxx` can you give me a referral to `bar-KDC`
7. foo-KDC: Yep. <returns `xxx-from-foo-KDC`> --->
8. Dear bar-KDC, I am `xxx-from-foo-KDC` do you have `aaa`?
9. bar-KDC: Yep!

Step 2 in the first block triggers lines 653-686, which is the KDC telling the client to go look elsewhere. 

Since we don't have a KDC telling us this, we're explicitly priming the next iteration to ask for the equivalent of a referral, that way it eventually goes through 653-686 albeit after an additional iteration, and from then on behaves just as the first block did.

It's a little chattier since it has one additional roundtrip, but that's the penalty all clients must pay if the KDC can't tell them where to go (Active Directory solves this by keeping various mappings of known names, but some companies don't deploy that particular feature).

Reply:
thanks for explanation. I missed the loop. It was off my diff but the "continue" was probably good enough clue I missed.